Adding two IntelliJ tests

anchouls opened the pull request:
`getMethodsContainingContent` - returns the name of the method that contains the given content

`existsConstantWithTheValue` - returns true if the file contains a constant with the given value

onewhl commented on ij/src/main/kotlin/org/jetbrains/academy/test/system/test/BaseIjTestClass.kt:
```diff
@@ -1,8 +1,46 @@
+        ApplicationManager.getApplication().runReadAction<Unit?> {
+            myFixture.file.accept(object : PsiRecursiveElementWalkingVisitor() {
+                override fun visitElement(element: PsiElement) {
+                    if (element is KtNamedFunction) currentMethod = element.name.toString()
```
What do you think about using `when` here?

onewhl commented on ij/src/main/kotlin/org/jetbrains/academy/test/system/test/BaseIjTestClass.kt:
```diff
@@ -1,8 +1,46 @@
-open class BaseIjTestClass : BasePlatformTestCase()
+open class BaseIjTestClass : BasePlatformTestCase() {
+
+    fun getMethodsContainingContent(content: String): List<String> {
```
What do you think about `findMethodsWithContent` name?

onewhl commented on ij/src/main/kotlin/org/jetbrains/academy/test/system/test/BaseIjTestClass.kt:
```diff
@@ -1,8 +1,46 @@
+            myFixture.file.accept(object : PsiRecursiveElementWalkingVisitor() {
+                override fun visitElement(element: PsiElement) {
+                    if (element is KtProperty && element.modifierList?.text?.contains("const") == true) {
+                        if (element.text.contains(elementValue)) existsConstant = true
```
It seems to me that we will walk through all const properties not depending if we already found the needed one. Can we stop walking when we found the needed const property?

onewhl commented on ij/src/main/kotlin/org/jetbrains/academy/test/system/test/BaseIjTestClass.kt:
```diff
@@ -1,8 +1,46 @@
+        return methodNames
+    }
+
+    fun existsConstantWithTheValue(elementValue: String): Boolean {
```
What do you think about `hasConstantWithGivenValue` name?

onewhl reviewed (changes requested):
@anchouls thank you! Please, take a look at the comments.

onewhl commented:
@anchouls to detect Extract Method refactoring, I would also check that there was added a call of newly added method inside the old one. The algorithm is described below:
1. Check that a student added a new method.
2. Check that the content of new method _partly_ matches with content of any others methods BEFORE (in the provided code before any student's changes).
3. Check that the old method has a call of new method.
4. Check that the content (body of new method) was removed from the old method.

What do you think? Do we need to implement it in test framework or in refactoring-course plugin?

anchouls commented:
@onewhl  
The algorithm that I implemented in the Extract method Test is like this:
1. Find the name of the method where the student has extracted the duplicated code. 
2. Check that there is only one such method - it means that the student removed the duplicate and added a new method. 
3. Look for methods where the new method is called. Check that these methods correspond to the necessary ones. (i.e. check that the old methods have a call to the new method).

It seems that this algorithm checks all the checks from the algorithm you described.

This algorithm includes two calls `findMethodsWithContent` method and `Assertions`, so I think it is enough to put only the `findMethodsWithContent` method into the test framework

nbirillo commented on ij/src/main/kotlin/org/jetbrains/academy/test/system/test/BaseIjTestClass.kt:
```diff
@@ -1,8 +1,55 @@
+        val methodNames = mutableListOf<String>()
+        var currentMethod = ""
+        ApplicationManager.getApplication().runReadAction<Unit?> {
+            myFixture.file.accept(object : PsiRecursiveElementWalkingVisitor() {
+                override fun visitElement(element: PsiElement) {
+                    when (element) {
+                        is KtNamedFunction -> currentMethod = element.name.toString()
+                        is KtBlockExpression -> {
+                            if (element.text.contains(content)) {
+                                methodNames.add(currentMethod)
+                            }
+                        }
+                    }
+                    super.visitElement(element)
+                }
+            })
+        }
```
I think it is better here to collect all KtNamedFunction elements and then to accepts a filter.

```kotlin
/** Extracts [kotlin elements][KtElement] of given type from kotlin related files in project. */
/** Extracts elements of given type from [PsiElement] subtree. */
fun <T : PsiElement> PsiElement.extractElementsOfType(psiElementClass: Class<T>): MutableCollection<T> =
    PsiTreeUtil.collectElementsOfType(this, psiElementClass)

myFixture.file.extractElementsOfType(KtNamedFunction::class.java)
```

Because in your case you can catch some troubles with nested functions

nbirillo commented on ij/src/main/kotlin/org/jetbrains/academy/test/system/test/BaseIjTestClass.kt:
```diff
@@ -1,8 +1,55 @@
+        var existsConstant = false
+        ApplicationManager.getApplication().runReadAction<Unit?> {
+            myFixture.file.accept(object : PsiRecursiveElementWalkingVisitor() {
+                override fun visitElement(element: PsiElement) {
```
You can reuse `extractElementsOfType` here too.

```kotlin
    private fun KtProperty.getConstValue(): String? {
        val possibleValue = extractElementsOfType(KtConstantExpression::class.java)
        if (possibleValue.isEmpty()) {
            return null
        }
        require(possibleValue.size == 1) { "Parser error! A const variable must have only one value" }
        return possibleValue.first().text.trimIndent()
    }

    fun hasConstantWithGivenValue(elementValue: String) = ApplicationManager.getApplication().runReadAction<Boolean> {
        val elements =  myFixture.file.extractElementsOfType(KtProperty::class.java)
        val a = elements.first().children
        elements.any { it.modifierList?.text?.contains("const") ?: false && it.getConstValue() == elementValue }
    }
```

nbirillo replied:
`element.text.contains(elementValue)` does not work since the variable name can contains elementValue

nbirillo commented on ij/src/main/kotlin/org/jetbrains/academy/test/system/test/BaseIjTestClass.kt:
```diff
@@ -1,8 +1,55 @@
-open class BaseIjTestClass : BasePlatformTestCase()
+open class BaseIjTestClass : BasePlatformTestCase() {
+
+    fun findMethodsWithContent(content: String): List<String> {
```
It does not work, students can use different formatting etc. You need to create a psi by content and use CodeFormatter to reformat code, Also users can create nested functions and in this case `element.text.contains(content)` does not work, you need to separate the function body - you can find the body (like I showed in getConstValue function) and compare it directly

nbirillo replied:
Also I'd suggest to move these functions into src and here just call them with fixture. Also, please ad descriptions

anchouls replied:
Below you suggest putting it in a new class that implements BaseIjTestClass - https://github.com/jetbrains-academy/kotlin-test-framework/pull/18#discussion_r1269489699. It seems to conflict with this comment. What did you mean?

nbirillo commented on ij/src/test/kotlin/org/jetbrains/academy/test/system/ij/formatting/BaseIjTestClassTests.kt:
```diff
@@ -0,0 +1,44 @@
+        """.trimIndent()
+        myFixture.configureByText("Task.kt", example)
+        val content = "println(\"Content\")"
+        assert(listOf("method1", "method3").equals(findMethodsWithContent(content)))
```
Please, put error messages to all asserts

nbirillo commented on ij/src/test/kotlin/org/jetbrains/academy/test/system/ij/formatting/BaseIjTestClassTests.kt:
```diff
@@ -0,0 +1,44 @@
+class BaseIjTestClassTests : BaseIjTestClass() {
+
+    fun testFindMethodsWithContent() {
+        val example = """
```
Add methods with nested bodies as an extra test case, please

nbirillo commented on ij/src/test/kotlin/org/jetbrains/academy/test/system/ij/formatting/BaseIjTestClassTests.kt:
```diff
@@ -0,0 +1,44 @@
+
+    fun testHasConstantWithGivenValue() {
+        val example = """
+            private const val CONSTANT1 = "some text"
```
Add const value into const name as an extra test case, please

nbirillo replied:
Also add different tests cases with broken formatting

nbirillo commented on ij/src/test/kotlin/org/jetbrains/academy/test/system/ij/formatting/BaseIjTestClassTests.kt:
```diff
@@ -0,0 +1,44 @@
+        myFixture.configureByText("Task.kt", example)
+        assert(hasConstantWithGivenValue("\"some text\""))
+        assert(hasConstantWithGivenValue("2"))
+        assert(!hasConstantWithGivenValue("0.5"))
```
You can use `assertFalse`

nbirillo commented on ij/src/main/kotlin/org/jetbrains/academy/test/system/test/BaseIjTestClass.kt:
```diff
@@ -1,8 +1,55 @@
  * A Base test class to create tests with PSI without adding intellij dependency into courses directly.
  */
-open class BaseIjTestClass : BasePlatformTestCase()
+open class BaseIjTestClass : BasePlatformTestCase() {
```
Probably you need to create a new class with a specific goal that implements BaseIjTestClass, isn't? Because it is impossible to add everything to BaseIjTestClass class

nbirillo reviewed (changes requested):
Thanks

nbirillo commented on ij/src/main/kotlin/org/jetbrains/academy/test/system/ij/analyzer/IjCodeAnalyzerUtil.kt:
```diff
@@ -0,0 +1,81 @@
+
+/** Extracts [kotlin elements][KtElement] of given type from kotlin related files in project. */
+/** Extracts elements of given type from [PsiElement] subtree. */
+fun <T : PsiElement> PsiElement.extractElementsOfTypes(psiElementClassList: List<Class<out T>>): MutableCollection<T> =
```
I'd prefer to change  `psiElementClassList: List<Class<out T>>` to `vararg`, it will help to avoid creating lists with one argument during calling this function, what do you think?

nbirillo reviewed (changes requested):
Only one small comment left, thanks!
